src/main/java/org/junit$pioneer/jupiter/TempDirectory.java

```diff
@@ -0,0 +1,245 @@
+ * via {@link org.junit.jupiter.api.extension.RegisterExtension @RegisterExtension}
+ * to pass a custom provider to configure the parent directory for all temporary
+ * directories created by this extension. This allows the use of this extension
+ * with any third-party {@code FileSystem} implementation, e.g.
+ * <a href="https://github.com/google/jimfs">Jimfs</a>.
+ *
+ * @since 0.1
+ * @see TempDir
+ * @see ParentDirProvider
+ * @see Files#createTempDirectory
+ */
+public class TempDirectory implements ParameterResolver {
+
+	/**
+	 * {@code TempDir} can be used to annotate a test or lifecycle method or
+	 * test class constructor parameter of type {@link Path} that should be
+	 * resolved into a temporary directory.
+	 *
+	 * @see TempDirectory
+	 */
+	@Target(ElementType.PARAMETER)
+	@Retention(RetentionPolicy.RUNTIME)
+	@Documented
+	public @interface TempDir {
+	}
+
+	/**
+	 * {@code ParentDirProvider} can be used to configure a custom parent
+	 * directory for all temporary directories created by the
+	 * {@link TempDirectory} extension this is used with.
+	 *
+	 * @see org.junit.jupiter.api.extension.RegisterExtension
+	 * @see TempDirectory#TempDirectory(ParentDirProvider)
+	 */
+	@FunctionalInterface
+	public interface ParentDirProvider {
+		/**
+		 * Get the parent directory for all temporary directories created by the
+		 * {@link TempDirectory} extension this is used with.
+		 *
+		 * @return the parent directory for all temporary directories; must be
+		 * {@code null} or an existing directory.
+		 */
+		Path get(ParameterContext parameterContext, ExtensionContext extensionContext) throws Exception;
+	}
+
+	private static final Namespace NAMESPACE = Namespace.create(TempDirectory.class);
+	private static final String KEY = "temp.dir";
+
+	private final ParentDirProvider parentDirProvider;
+
+	/**
+	 * Create a new {@code TempDirectory} extension that uses the default
+	 * {@link java.nio.file.FileSystem FileSystem} and creates temporary
+	 * directories in the default location.
+	 *
+	 * <p>This constructor is used by the JUnit Jupiter Engine when the
+	 * extension is registered via
+	 * {@link org.junit.jupiter.api.extension.ExtendWith @ExtendWith}.
+	 */
+	public TempDirectory() {
+		this((parameterContext, extensionContext) -> null);
+	}
+
+	/**
+	 * Create a new {@code TempDirectory} extension that uses the supplied
+	 * {@link Callable} to configure the parent directory for the temporary
+	 * directories created by this extension.
+	 *
+	 * <p>The parent directory returned by the supplied {@link Callable} must
+	 * be {@code null} or an existing directory.
+	 *
+	 * <p>You may use this constructor when registering this extension via
+	 * {@link org.junit.jupiter.api.extension.RegisterExtension @RegisterExtension}.
+	 *
+	 * @param parentDirProvider used to configure the parent directory for the
+	 * temporary directories created by this extension; never {@code null}
```

Comment:
You can remove `never {@code null}` - in Pioneer nothing is ever allowed to be `null`.